server/src/main/java/org/elasticsearch/index/codec/vectors/DefaultIVFVectorsWriter.java

```diff
@@ -302,10 +276,28 @@ CentroidAssignments calculateAndWriteCentroids(
             infoStream.message(IVF_VECTOR_COMPONENT, "final centroid count: " + centroids.length);
         }
 
+        IntArrayList[] assignmentsByCluster = new IntArrayList[centroids.length];
+        for (int c = 0; c < centroids.length; c++) {
+            IntArrayList cluster = new IntArrayList(vectorPerCluster);
+            for (int j = 0; j < assignments.length; j++) {
+                if (assignments[j] == c) {
+                    cluster.add(j);
+                }
+            }
+
+            for (int j = 0; j < soarAssignments.length; j++) {
+                if (soarAssignments[j] == c) {
+                    cluster.add(j);
+                }
+            }
+
+            assignmentsByCluster[c] = cluster;
+        }
```

Comment:
```suggestion
        if (soarAssignments.length != assignments.length) {
            throw new IllegalStateException(
                "soar assignments length does not match assignments length: " + soarAssignments.length + " != " + assignments.length
            );
        }
        for (int j = 0; j < assignments.length; j++) {
            if (assignmentsByCluster[assignments[j]] == null) {
                assignmentsByCluster[assignments[j]] = new IntArrayList(vectorPerCluster);
            }
            if (assignmentsByCluster[soarAssignments[j]] == null) {
                assignmentsByCluster[soarAssignments[j]] = new IntArrayList(vectorPerCluster);
            }
            assignmentsByCluster[assignments[j]].add(j);
            assignmentsByCluster[soarAssignments[j]].add(j);
        }
```
What do you think of this? 

Instead of doing `soarAssignments.length * centroids.length +  assignments.length * centroids.length` iterations, we can do it in a single loop. We always overspill right now and we can initialize the centroid list if it doesn't exist.

Reply:
hmm, this breaks all of the IVFVectorFormatTests; let me see if I can easily identify why

Reply:
oh oh that assumption in the if clause won't always be valid.  It's using this logic for both flush and merge now and on flush there are no soar assignments.  I'll keep the code as is unless you have a better suggestion here.